### DISCUSSION TITLE
the gpu output results are different from cpu when I used mxnet-cu100 in adversarial_action_recognition.ipynb

### DISCUSSION THREAD

**Discussion by AndssY** [Trusted-AI/adversarial-robustness-toolbox#1578] [Q&A] [answered]
When I started running this script(adversarial_action_recognition.ipynb) with GPU, the ART classifier outputs are right, but the adversarial attack outputs become: 
![image](https://user-images.githubusercontent.com/11785282/157441584-25e3902e-b533-4ceb-9537-f0ab983d3b2f.png)
which are different from the CPU version which in notebooks:
![image](https://user-images.githubusercontent.com/11785282/157443865-d4f6cdaf-8ad2-42ce-8126-68b34b6b3d98.png)

I'm a green hand, do not well in programming. I want to know, can I run this script with GPU to speed up? And if can, how to change it to run it correctly?

My changes are all listed below(except these, I didn't change anything else):
```
nd.array()    ->    nd.array(ctx=mx.gpu())    #in every nd.array()
...
model=get_model()    ->    model = get_model(ctx=mx.gpu())
....
classifier_art = MXClassifier()    ->    classifier_art = MXClassifier(ctx=mx.gpu())
```

my system information are: 
OS: Windows10
Python version: 3.7.0
ART version: 1.9.1
MXNet version: mxnet-cu100==1.5.0

**Comment by beat-buesser** [comment, marked as answer]
Hi @AndssY Thank you for using ART! Theoretically, the results from CPU and GPU runs should be very similar. What is the maximal absolute difference between `adv_sample` and `adv_sample_input` after cell 10 for CPU and GPU runs?

**Reply by AndssY** [reply]
Thanks for help! When I trying to caculate the difference between `adv_sample` and `adv_sample_input`
by `print(max(adv_sample-adv_sample_input))`, I found the maximal absolute difference is about `0.0313`(values are `0.0313725`, `-0.0313725` or `0`) for both GPU and CPU runs. 
It seemed OK, But the adversarial attack result is different from the theory.

**Reply by beat-buesser** [reply]
@AndssY Ok, these numbers are good. Would you be able to upgrade mxnet to 1.8 and try again?

**Reply by AndssY** [reply]
I‘m sorry but I can't find the mxnet-cu10x==1.8 for windows10. I used to look for it, but It seemed that 1.8 does not have windows version
`https://dist.mxnet.io/python` there may show mxnet1.8 for GPU only has linux version. perhaps I will try to upgrade to 1.9 or 2.0

**Reply by beat-buesser** [reply]
I have been able to reproduce the issue with 1.7 on Linux.

**Reply by beat-buesser** [reply]
It looks like the gradients returned by `loss_gradient` are `nan` if `ctx=mx.gpu()`.

**Reply by beat-buesser** [reply]
One thing I noticed, the loss should be from logits: `loss=gluon.loss.SoftmaxCrossEntropyLoss(from_logits=True),`, but it didn't solve the issue.

Repeating calls `loss_gradient` leads occasionally to non-nan gradients, but mostly the behavior seems similar as described also here: https://github.com/awslabs/gluon-ts/issues/1904

**Reply by AndssY** [reply]
I have been meet this situation(`loss_gradient` are `nan`) when I use this script to attack other example in UCF101.

after trying `loss=gluon.loss.SoftmaxCrossEntropyLoss(from_logits=True)`, the result becomes:
![image](https://user-images.githubusercontent.com/11785282/157788211-532f5a66-750b-4df3-bd06-26f2b433ed60.png)

So is this problem related to Mxnet and now we can't know what happen and it can't be fixed?
(why output are different, why NaN gradient, when use GPU but works fine for CPU)

Would it be better to replace the deep learning framework with PyTorch or TensorFlow in `adversarial_action_recognition.ipynb`? 
or I want to know how to replace Mxnet to other framework(like PyTorch), is there some tutorials here? or what can I learn to attack a action recognition model in PyTorch? : ) Thanks again for help!

**Reply by beat-buesser** [reply]
Reproducing the notebook in another deep learning framework is definitely possible. The two main steps you would have to change are getting a model for action recognition in another framework and replace `MXClassifier` with `PyTorchClassifier`, for example.
